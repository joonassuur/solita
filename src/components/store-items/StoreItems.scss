@import '../../variables.scss';

#cart-content .storeItem {
  border-bottom: 1px solid black;
}
#product-price {
  font-weight: 400;
  text-align: right;
}
#product-quantity {
  text-align: right;
  font-weight: 400;
  margin-bottom: 0.5em;
}
#product-name {
  font-style: italic;
}
#addRemove-btn {
  margin-top: 0.5em;
}
.storeItem {
  display: flex;
  padding: 2em 0;
  justify-content: space-between;
  .left,
  .right {
    align-items: center;
    display: flex;
  }
  .image {
    position: relative;
    display: flex;
    align-items: center;
    left: 0;
    height: 80px;
    width: 80px;
    margin-right: 2em;
    &.triangle:before {
      content: '';
      position: absolute;
      top: 0;
      left: 50%;
      transform: translateX(-50%);
      border-left: 40px solid transparent;
      border-right: 40px solid transparent;
      border-bottom: 80px solid palevioletred;
    }
    &.triangle:after {
      content: '';
      position: absolute;
      top: 3px;
      left: 50%;
      transform: translateX(-50%);
      border-left: 37px solid transparent;
      border-right: 37px solid transparent;
      border-bottom: 75px solid #fff;
    }
    &.rectangle:before {
      content: '';
      position: absolute;
      top: 0;
      left: 50%;
      transform: translateX(-50%);
      border-left: 40px solid palevioletred;
      border-right: 40px solid palevioletred;
      border-bottom: 80px solid palevioletred;
    }
    &.rectangle:after {
      content: '';
      position: absolute;
      top: 2px;
      left: 50%;
      transform: translateX(-50%);
      border-left: 38px solid white;
      border-right: 38px solid white;
      border-bottom: 76px solid white;
    }
    &.circle:before {
      content: '';
      border-radius: 100px;
      position: absolute;
      top: 0;
      left: 50%;
      transform: translateX(-50%);
      border-left: 40px solid palevioletred;
      border-right: 40px solid palevioletred;
      border-bottom: 40px solid palevioletred;
      border-top: 40px solid palevioletred;
    }
    &.circle:after {
      content: '';
      border-radius: 100px;
      position: absolute;
      top: 2px;
      left: 50%;
      transform: translateX(-50%);
      border-left: 38px solid white;
      border-right: 38px solid white;
      border-bottom: 38px solid white;
      border-top: 38px solid white;
    }
  }
}

@media (max-width: $md) {
  .storeItem {
    flex-direction: column;
    .image {
      display: none;
    }
    #product-price {
      margin-top: .5em;
    }
    #product-price,
    #product-quantity {
      text-align: left;
    }
  }
}
