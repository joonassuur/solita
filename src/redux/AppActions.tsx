import { createAction, createAsyncThunk } from "@reduxjs/toolkit";
import products from "../mockedAPI/products.json";

export const fetchStoreData = createAsyncThunk(
  "@app: fetchStoreData",
  async () => {
    // fake the api call
    const res = products;
    return res;
  }
);
export const modifyCart = createAction("@app: modifyCart", (item) => {
  return { payload: item };
});
export const toggleModal = createAction("@app: toggleModal", (isOpen) => {
  return { payload: isOpen };
});
