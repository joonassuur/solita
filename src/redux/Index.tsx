export {
  fetchStoreData,
  modifyCart,
} from "./AppActions";

export { getProducts, getCart } from "./Selectors";
