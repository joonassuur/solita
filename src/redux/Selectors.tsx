import { StoreItem } from "../types/Types";
export const getProducts = (state: { app: { products: StoreItem[] } }) =>
  state.app.products;
export const getCart = (state: { app: { cart: StoreItem[] } }) =>
  state.app.cart;
