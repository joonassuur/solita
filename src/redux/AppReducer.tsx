import { createSlice } from "@reduxjs/toolkit";
import { fetchStoreData, modifyCart, toggleModal } from "../redux/AppActions";
import { Store } from "../types/Types";

import _ from "lodash";

const initialState: Store = {
  products: [],
  cart: [],
  isModalOpen: false,
};

const app = createSlice({
  name: "app",
  initialState,
  reducers: {},
  extraReducers: {
    [fetchStoreData.fulfilled.toString()]: (state, { payload }) => {
      state.products = payload.products;
    },
    [fetchStoreData.rejected.toString()]: (state) => {
      return state;
    },
    [modifyCart.toString()]: (state, { payload }) => {
      const clonedCart = _.cloneDeep(state.cart);
      const existingItemFilter = clonedCart.filter(
        (cartItem) => cartItem.id === payload.id
      );
      
      if (existingItemFilter.length > 0) {
        clonedCart.map((cartItem) =>
          cartItem.id === existingItemFilter[0].id
            ? (cartItem.quantity = payload.quantity)
            : false
        );
      } else {
        clonedCart.push(payload);
      }

      state.cart = clonedCart;
    },
    [toggleModal.toString()]: (state, { payload }) => {
      state.isModalOpen = payload;
    },
  },
});

export default app;
