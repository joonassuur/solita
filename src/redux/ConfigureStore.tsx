import { createStore, applyMiddleware  } from "redux";
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';

import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import rootReducer from "./RootReducer";

const persistConfig = {
  key: "root",
  storage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export let store = createStore(persistedReducer, composeWithDevTools(applyMiddleware(thunk)));
export let persistor = persistStore(store);
