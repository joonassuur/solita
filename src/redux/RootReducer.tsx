import { combineReducers } from "redux";
import app from "./AppReducer";

const rootReducer = combineReducers({
  app: app.reducer,
});
export default rootReducer;
