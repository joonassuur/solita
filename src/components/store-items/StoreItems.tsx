import React from "react";
import { useDispatch } from "react-redux";
import { addToCart, removeFromCart } from "../../redux/Index";
import { StoreItem } from "../../types/Types";

import "./StoreItems.scss";

function StoreItems(products: StoreItem[], cart: StoreItem[]) {

  const dispatch = useDispatch();

  const addOrRemoveItem = (i: StoreItem) => {
    // if item exists in cart, remove it, otherwise add to cart
    if (cart.some((x: StoreItem) => x.id === i.id)) {
      dispatch(removeFromCart(i));
    } else {
      dispatch(addToCart(i));
    }
  };
  
  return (
    <>
      {products.map((e: StoreItem) => {
        return (
          <div key={e.id} className="storeItem">
            <div className="left">
              <div className={`image ${e.name.toLowerCase()}`}></div>
              <div className="name-desc">
                <h1 id="product-name">{e.name}</h1>
                <div>{e.description}</div>
              </div>
            </div>
            <div className="right">
              <div className="price-btn">
                <h1 id="product-price">{`${e.price} â‚¬`}</h1>
                <button onClick={() => addOrRemoveItem(e)}>
                  {cart.some((x: StoreItem) => x.id === e.id)
                    ? "Remove from cart"
                    : "Add to cart"}
                </button>
              </div>
            </div>
          </div>
        );
      })}
    </>
  );
}

export default StoreItems;