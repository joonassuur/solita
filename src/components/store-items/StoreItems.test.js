import React from "react";
import { shallow } from "enzyme";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import StoreItems from "./StoreItems";
import { getCart, getProducts } from "../../redux/Selectors";
import { addToCart, removeFromCart } from "../../redux/AppActions";
import { products } from "../../mockedAPI/products.json";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const mockDispatch = jest.fn();

jest.mock("react-redux", () => ({
  useSelector: jest.fn(fn => fn()),
  useDispatch: () => mockDispatch
}));
jest.mock("../../redux/Selectors");
getCart.mockReturnValue([]);
getProducts.mockReturnValue(products);

describe("StoreItems Component", () => {
  // Initialize mockstore with empty state
  let initialState;
  let store;
  beforeEach(() => {
    initialState = {
      products: [],
      cart: []
    };
    store = mockStore(initialState)
  });

  it("should dispatch addToCart action", () => {
    // Dispatch the action
    store.dispatch(addToCart());

    // Test if your store dispatched the expected actions
    const actions = store.getActions();
    const expectedPayload = { type: "@app: addToCart" };
    expect(actions).toEqual([expectedPayload]);
  });

  it("should dispatch removeFromCart action", () => {
    // Dispatch the action
    store.dispatch(removeFromCart());

    // Test if your store dispatched the expected actions
    const actions = store.getActions();
    const expectedPayload = { type: "@app: removeFromCart" };
    expect(actions).toEqual([expectedPayload]);
  });

  it("Should match snapshot", () => {
    const props = {
      cartAction: "add",
      buttonText: "Add to Cart",
      renderElement: "products"
    };
    const testRenderer = shallow(<StoreItems {...props} />);
    expect(testRenderer).toMatchSnapshot();
  });

  it("Should match snapshot", () => {
    const props = {
      cartAction: "remove",
      buttonText: "Remove from Cart",
      renderElement: "cart"
    };
    const testRenderer = shallow(<StoreItems {...props} />);
    expect(testRenderer).toMatchSnapshot();
  });
});
