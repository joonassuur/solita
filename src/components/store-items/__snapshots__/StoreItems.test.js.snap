// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreItems Component Should match snapshot 1`] = `ShallowWrapper {}`;

exports[`StoreItems Component Should match snapshot 2`] = `ShallowWrapper {}`;
